File: dep/guifast_shared/frontend/src/action/forward_action.ts
import { Action as ActionBase, ActionLocation, actionStrId, String } from "guifast_shared";

export namespace ForwardAction {
    export const str = "forward_action";
    export const id = actionStrId(String.module, str);

    export interface Action extends ActionBase {
        readonly action: ActionBase,
        readonly destination: ActionLocation,
    }

    export const make = (
        action: ActionBase,
        destination: ActionLocation
    ): Action => {
        return {
            type: id,
            action: action,
            destination: destination
        };
    };
}

File: dep/guifast_shared/backend/src/action/forward_action.rs
use libflo_action_std::{ Action, action_str_id, NumberOrString };
use serde::Serialize;
use serialization::ActionLocation;
use string;

#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
pub struct ForwardAction<'a, T>
    where T: Action + Serialize + 'a {
    #[serde(rename = "type")]
    action_type: NumberOrString,
    action: &'a T,
    destination: ActionLocation,
}

impl <'a, T> ForwardAction<'a, T>
    where T: Action + Serialize + 'a {
    pub fn new(action: &'a T, destination: ActionLocation) -> Self {
        ForwardAction {
            action_type: action_str_id(string::module(), string::forward_action()),
            action: action,
            destination: destination
        }
    }
}

impl <'a, T> Action for ForwardAction<'a, T>
    where T: Action + Serialize + 'a {
    fn get_type(&self) -> &NumberOrString {
        &self.action_type
    }
}
